core/src/eval/cache/lazy.rs

```diff
@@ -647,18 +681,21 @@ impl Cache for CBNCache {
         &mut self,
         idx: &mut CacheIndex,
     ) -> Result<Option<Self::UpdateIndex>, BlackholedError> {
-        if idx.state() != ThunkState::Evaluated {
-            if idx.should_update() {
-                idx.mk_update_frame().map(Some)
-            }
-            // If the thunk isn't to be updated, directly set the evaluated flag.
-            else {
-                idx.set_evaluated();
-                Ok(None)
-            }
-        } else {
-            Ok(None)
+        // If the thunk is already evaluated, we don't return an update frame.
```

Comment:
The changes are a bit cosmetic, but not only: previously, if `idx.should_update()` was `false`, we would never look at the state of the thunk and always proceed with evaluation. This is actually not correct, because if the thunk is blackholed, we should raise an infinite recursion error. For eval record spine in particular, the parents thunk are in weak head normal form (they are records), so `should_update()` is false and without this additional change, the detection of the infinite cycle fails.

Reply:
I guess this code path is no longer triggered in `eval_record_spine`? But maybe it makes sense to do this anyway...

Reply:
Yes, I suspect the previous code could miss some infinite recursion errors as well because of that.